scd/usbtoken.c

```diff
@@ -0,0 +1,737 @@
+	}
+
+	/* after pairing (pin should no longer be in transport state), we generate a
+	 * master secret from which the subsequent keys can be derived.
+	 * Device initialization must only be possible for user in possession of SO-PIN
+	 * TODO: document security arch and workflow
+	 */
+	rc = generateSymmetricKey(token->ctn, 1, algo_dskkey, sizeof(algo_dskkey));
+	if (rc < 0) {
+		ERROR("USBTOKEN: generateSymmetricKey() failed with code: %d", rc);
+		rc = -1;
+		goto out;
+	}
+
+	rc = writeKeyDescription(token->ctn, 1, skd_dskkey, sizeof(skd_dskkey));
+	if (rc < 0) {
+		ERROR("USBTOKEN: writeKeyDescription() failed with code: %d", rc);
+		rc = -1;
+	}
+
+out:
+	CT_close(token->ctn);
+	return rc;
+}
+
+static int
+change_user_pin(usbtoken_t *token, const char *oldpass, const char *newpass,
+		unsigned char *pairing_secret, size_t pairing_sec_len)
+{
+	TRACE("USBTOKEN: change_user_pin");
+
+	ASSERT(token);
+	ASSERT(oldpass);
+	ASSERT(newpass);
+	ASSERT(pairing_secret);
+
+	int rc;
+	int oldcode_len, newcode_len;
+	unsigned char oldcode[TOKEN_MAX_AUTH_CODE_LEN];
+	unsigned char newcode[TOKEN_MAX_AUTH_CODE_LEN];
+
+	rc = usbtoken_init_ctapi_int(token);
+	if (rc != 0) {
+		ERROR("Failed to initialize ctapi interface to usb token reader");
+		return -1;
+	}
+
+	newcode_len =
+		get_auth_code(newpass, pairing_secret, pairing_sec_len, newcode, sizeof(newcode));
+	if (newcode_len < 0) {
+		ERROR("Could not derive new authentication code");
+		rc = -1;
+		goto out;
+	}
+
+	oldcode_len =
+		get_auth_code(oldpass, pairing_secret, pairing_sec_len, oldcode, sizeof(oldcode));
+	if (oldcode_len < 0) {
+		ERROR("Could not derive old authentication code");
+		rc = -1;
+		goto out;
+	}
+
+	rc = changePIN(token->ctn, oldcode, oldcode_len, newcode, newcode_len);
+
+	memset(newcode, 0, sizeof(newcode));
+	memset(oldcode, 0, sizeof(oldcode));
+
+	if (rc != 0x9000) {
+		ERROR("Changing user pin of usbtoken failed. Token rc: 0x%04x", rc);
+		rc = -1;
+	} else
+		rc = 0;
+
+out:
+	CT_close(token->ctn);
+	return rc;
+}
+
+/** scd interface to usbtoken **/
+
+int
+usbtoken_change_passphrase(usbtoken_t *token, const char *oldpass, const char *newpass,
+			   unsigned char *pairing_secret, size_t pairing_sec_len,
+			   bool is_provisioning)
+{
+	TRACE("USBTOKEN: usbtoken_change_passphrase");
+
+	ASSERT(token);
+	ASSERT(oldpass);
+	ASSERT(newpass);
+
+	return (is_provisioning ?
+			provision_auth_code(token, oldpass, newpass, pairing_secret,
+					    pairing_sec_len) :
+			change_user_pin(token, oldpass, newpass, pairing_secret, pairing_sec_len));
+}
+
+/**
+ * Free secrets.
+ */
+static void
+usbtoken_free_secrets(usbtoken_t *token)
+{
+	TRACE("USBTOKEN: usbtoken_free_secrets");
+
+	ASSERT(token);
+
+	memset(token->auth_code, 0, token->auth_code_len);
+	mem_free(token->auth_code);
+}
+
+void
+usbtoken_free(usbtoken_t *token)
+{
+	TRACE("USBTOKEN: usbtoken_free");
+
+	ASSERT(token);
+
+	mem_free(token->serial);
+	usbtoken_free_secrets(token);
+
+	mem_free(token);
+	g_ctn--;
```

Comment:
better don't reuse ctn => just increment (add overflow check)

Reply:
seemed to work in my test. i don't like the idea of having  "zombie" card terminal numbers (although the range should be large enough even for systems that are up for a long time)